README.md
# <img alt="@bull-board" src="https://raw.githubusercontent.com/felixmosh/bull-board/master/packages/ui/src/static/images/logo.svg" width="35px" /> @bull-board

Bull Dashboard is a UI built on top of [Bull](https://github.com/OptimalBits/bull) or [BullMQ](https://github.com/taskforcesh/bullmq) to help you visualize your queues and their jobs.
With this library you get a beautiful UI for visualizing what's happening with each job in your queues, their status and some actions that will enable you to get the job done.

<p align="center">
  <a href="https://www.npmjs.com/org/bull-board">
    <img alt="npm downloads" src="https://img.shields.io/npm/dw/@leo-guinan/api">
  </a>
  <a href="https://github.com/felixmosh/bull-board/blob/master/LICENSE">
    <img alt="licence" src="https://img.shields.io/github/license/felixmosh/bull-board">
  </a>
  <img alt="open issues" src="https://img.shields.io/github/issues/felixmosh/bull-board"/>
<p>

![UI](https://raw.githubusercontent.com/felixmosh/bull-board/master/screenshots/dashboard.png)

## Packages

| Name                                                                     | Version                                                           |
|--------------------------------------------------------------------------|-------------------------------------------------------------------|
| [@leo-guinan/api](https://www.npmjs.com/package/@leo-guinan/api)         | ![npm (scoped)](https://img.shields.io/npm/v/@leo-guinan/api)     |
| [@leo-guinan/ui](https://www.npmjs.com/package/@leo-guinan/ui)           | ![npm (scoped)](https://img.shields.io/npm/v/@leo-guinan/ui)      |
| [@leo-guinan/express](https://www.npmjs.com/package/@leo-guinan/express) | ![npm (scoped)](https://img.shields.io/npm/v/@leo-guinan/express) |
| [@leo-guinan/fastify](https://www.npmjs.com/package/@leo-guinan/fastify) | ![npm (scoped)](https://img.shields.io/npm/v/@leo-guinan/fastify) |
| [@leo-guinan/koa](https://www.npmjs.com/package/@leo-guinan/koa)         | ![npm (scoped)](https://img.shields.io/npm/v/@leo-guinan/koa)     |
| [@leo-guinan/hapi](https://www.npmjs.com/package/@leo-guinan/hapi)       | ![npm (scoped)](https://img.shields.io/npm/v/@leo-guinan/hapi)    |

## Notes

As this library provides only the visualization for your queues, keep in mind that:

- You must have either [Bull](https://github.com/OptimalBits/bull) or [BullMQ](https://github.com/taskforcesh/bullmq) installed in your projects;
- Aside the options to retry and clean jobs, this library is not responsible for processing the jobs, reporting progress or any other thing. This must be done in your application with your own logic;
- If you want to understand the possibilities you have with the queues please refer to [Bull's docs](https://optimalbits.github.io/bull/) or [BullMQ's docs](https://docs.bullmq.io/);
- This library doesn't hijack Bull's way of working.

If you want to learn more about queues ([Bull](https://github.com/OptimalBits/bull) or [BullMQ](https://github.com/taskforcesh/bullmq)) and [Redis](https://redis.io/).

## Starting

To add it to your project start by installing a server framework specific adapter to your dependencies list:

```sh
yarn add @leo-guinan/express
# or
yarn add @leo-guinan/fastify
# or
yarn add @leo-guinan/hapi
# or
yarn add @leo-guinan/koa
```
Or
```sh
npm i @leo-guinan/express
# or
npm i @leo-guinan/fastify
# or
npm i @leo-guinan/hapi
# or
npm i @leo-guinan/koa
```

## Hello World

```js
const express = require('express');
const Queue = require('bull');
const QueueMQ = require('bullmq');
const { createBullBoard } = require('@leo-guinan/api');
const { BullAdapter } = require('@leo-guinan/api/bullAdapter');
const { BullMQAdapter } = require('@leo-guinan/api/bullMQAdapter');
const { ExpressAdapter } = require('@leo-guinan/express');

const someQueue = new Queue('someQueueName', {
  redis: { port: 6379, host: '127.0.0.1', password: 'foobared' },
}); // if you have a special connection to redis.
const someOtherQueue = new Queue('someOtherQueueName');
const queueMQ = new QueueMQ('queueMQName');

const serverAdapter = new ExpressAdapter();
serverAdapter.setBasePath('/admin/queues');

const { addQueue, removeQueue, setQueues, replaceQueues } = createBullBoard({
  queues: [new BullAdapter(someQueue), new BullAdapter(someOtherQueue), new BullMQAdapter(queueMQ)],
  serverAdapter: serverAdapter,
});

const app = express();

app.use('/admin/queues', serverAdapter.getRouter());

// other configurations of your server
  
app.listen(3000, () => {
  console.log('Running on 3000...');
  console.log('For the UI, open http://localhost:3000/admin/queues');
  console.log('Make sure Redis is running on port 6379 by default');
});
```

That's it! Now you can access the `/admin/queues` route, and you will be able to monitor everything that is happening in your queues 😁


For more advanced usages check the `examples` folder, currently it contains:
1. [Basic authentication example](https://github.com/felixmosh/bull-board/tree/master/examples/with-express-auth)
2. [Multiple instance of the board](https://github.com/felixmosh/bull-board/tree/master/examples/with-multiple-instances)
2. [With Fastify server](https://github.com/felixmosh/bull-board/tree/master/examples/with-fastify)
2. [With Hapi.js server](https://github.com/felixmosh/bull-board/tree/master/examples/with-hapi)
2. [With Koa.js server](https://github.com/felixmosh/bull-board/tree/master/examples/with-koa)
### Queue options
1. `readOnlyMode` (default: `false`)
Makes the UI as read only, hides all queue & job related actions

```js
const Queue = require('bull')
const QueueMQ = require('bullmq')
const { createBullBoard } = require('@leo-guinan/api')
const { BullMQAdapter } = require('@leo-guinan/api/bullMQAdapter')
const { BullAdapter } = require('@leo-guinan/api/bullAdapter')

const someQueue = new Queue()
const someOtherQueue = new Queue()
const queueMQ = new QueueMQ()

const { setQueues, replaceQueues } = createBullBoard({
  queues: [
    new BullAdapter(someQueue, { readOnlyMode: true }), // only this queue will be in read only mode
    new BullAdapter(someOtherQueue),
    new BullMQAdapter(queueMQ, { readOnlyMode: true }),
  ]
})
```

2. `allowRetries` (default: `true`)
When set to `false` the UI removes the job retry buttons for a queue. This option will be ignored if `readOnlyMode` is `true`.

```js
const QueueMQ = require('bullmq')
const { createBullBoard } = require('@leo-guinan/api')
const { BullMQAdapter } = require('@leo-guinan/api/bullMQAdapter')
const { BullAdapter } = require('@leo-guinan/api/bullAdapter')

const someQueue = new Queue()
const someOtherQueue = new Queue()
const queueMQ = new QueueMQ()

const { setQueues, replaceQueues } = createBullBoard({
  queues: [
    new BullAdapter(someQueue), 
    new BullAdapter(someOtherQueue, , { allowRetries: false }), // No retry buttons
    new BullMQAdapter(queueMQ, { allowRetries: true, readOnlyMode: true }), // allowRetries will be ignored in this case in lieu of readOnlyMode
  ]
})
```

### Hosting router on a sub path

If you host your express service on a different path than root (/) ie. https://<server_name>/<sub_path>/, then you can add the following code to provide the configuration to the bull-board router. In this example the sub path will be `my-base-path`.

```js
const Queue = require('bull')
const { createBullBoard } = require('@leo-guinan/api')
const { BullAdapter } = require('@leo-guinan/api/bullAdapter')
const { ExpressAdapter } = require('@leo-guinan/express')

const basePath = '/my-base-path';

const someQueue = new Queue('someQueueName')
const serverAdapter = new ExpressAdapter();
serverAdapter.setBasePath(basePath)

createBullBoard({
  queues: [
    new BullAdapter(someQueue),
  ],
  serverAdapter 
})

// ... express server configuration

app.use(basePath, serverAdapter.getRouter());
```

You will then find the bull-board UI at the following address `https://<server_name>/my-base-path/queues`.

## Contributing

First, thank you for being interested in helping out, your time is always appreciated in every way. 💯

Remember to read the [Code of Conduct](https://github.com/felixmosh/bull-board/blob/master/CODE_OF_CONDUCT.md) so you also help maintaining a good Open source community around this project!

Here are some tips:

- Check the [issues page](https://github.com/felixmosh/bull-board/issues) for already opened issues (or maybe even closed ones) that might already address your question/bug/feature request.
- When opening a bug report provide as much information as you can, some things might be useful for helping debugging and understading the problem
  - Node, Redis, Bull, BullMQ, bull-board versions
  - Sample code that reproduces the problem
  - Some of your environment details
  - Framework you're using (Express, Koa, Hapi, etc).
- Feature requests are welcomed! Provide some details on why it would be helpful for you and others, explain how you're using bull-board and if possible even some screenshots if you are willing to mock something!

## Developing

If you want to help us to solve the issues, be it a bug, a feature or a question, you might need to fork and clone this project.

To fork a project means you're going to have your own version of it under your own GitHub profile, you do it by clicking the "Fork" button on the top of any project's page on GitHub.

Cloning a project means downloading it to your local machine, you do it in the command line:

```sh
git clone git@github.com:YOUR_GITHUB_USERNAME/bull-board.git
```

That will create a `bull-board` folder inside the directory you executed the command, so you need to navigate inside it:

```sh
cd bull-board
```

_This project requires that you have [yarn](https://yarnpkg.com/lang/en/) installed_

Also make sure you are running **Redis** for this project (bull-board's example connects to Redis' default port `6379`).

Now, to try it out locally you can run:

```sh
yarn && yarn build && yarn start:dev
```

The ui open automaticlly in the browser at `http://localhost:3000/ui`

### Acknowledgements ❤️

- [Juan](https://github.com/joaomilho) for building the first version of this library

# License

This project is licensed under the [MIT License](https://github.com/felixmosh/bull-board/blob/master/LICENSE), so it means it's completely free to use and copy, but if you do fork this project with nice additions that we could have here, remember to send a PR 👍


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### 0.0.1

- feat: Log tab enhancements  [`#458`](https://github.com/leo-guinan/bull-board/pull/458)
- Add canRetry options to queue [`#384`](https://github.com/leo-guinan/bull-board/pull/384)
- chore(deps): bump follow-redirects from 1.14.7 to 1.14.8 [`#381`](https://github.com/leo-guinan/bull-board/pull/381)
- fix: decode queue name to support url special chars, closes #375 [`#376`](https://github.com/leo-guinan/bull-board/pull/376)
- chore(deps): bump node-fetch from 2.6.6 to 2.6.7 [`#372`](https://github.com/leo-guinan/bull-board/pull/372)
- Improve performance [`#371`](https://github.com/leo-guinan/bull-board/pull/371)
- chore(deps): bump nanoid from 3.1.30 to 3.2.0 [`#369`](https://github.com/leo-guinan/bull-board/pull/369)
- chore(deps): bump follow-redirects from 1.14.6 to 1.14.7 [`#365`](https://github.com/leo-guinan/bull-board/pull/365)
- feat: add job name formatter [`#359`](https://github.com/leo-guinan/bull-board/pull/359)
- chore(deps): bump postcss from 7.0.35 to 7.0.39 [`#356`](https://github.com/leo-guinan/bull-board/pull/356)
- chore(deps): bump fastify-static from 4.2.3 to 4.4.1 [`#354`](https://github.com/leo-guinan/bull-board/pull/354)
- chore(deps-dev): bump axios from 0.21.1 to 0.21.2 [`#353`](https://github.com/leo-guinan/bull-board/pull/353)
- chore(deps): bump tar from 4.4.13 to 4.4.19 [`#355`](https://github.com/leo-guinan/bull-board/pull/355)
- chore(deps): bump fastify-static in /examples/with-fastify-auth [`#352`](https://github.com/leo-guinan/bull-board/pull/352)
- add noindex meta tag [`#351`](https://github.com/leo-guinan/bull-board/pull/351)
- chore(deps): bump fastify-static in /examples/with-fastify [`#343`](https://github.com/leo-guinan/bull-board/pull/343)
- Add Hapi route options to plugin [`#342`](https://github.com/leo-guinan/bull-board/pull/342)
- chore(deps): bump fastify-static in /examples/with-fastify [`#337`](https://github.com/leo-guinan/bull-board/pull/337)
- feat: Add prefix option to allow multiple queues with the same name, but on different envs (prod, stage) [`#333`](https://github.com/leo-guinan/bull-board/pull/333)
- chore(deps): bump nth-check from 2.0.0 to 2.0.1 [`#328`](https://github.com/leo-guinan/bull-board/pull/328)
- Remove unused return value from `createBullBoard()` [`#329`](https://github.com/leo-guinan/bull-board/pull/329)
- fix: show retry button on failed tab (job can be retried) [`#327`](https://github.com/leo-guinan/bull-board/pull/327)
- chore(deps): bump tmpl from 1.0.4 to 1.0.5 [`#326`](https://github.com/leo-guinan/bull-board/pull/326)
- Rollback @radix-ui/react-dropdown-menu due to https://github.com/radix-ui/primitives/issues/837 [`#321`](https://github.com/leo-guinan/bull-board/pull/321)
- Move React development dependencies to devDependecies [`#317`](https://github.com/leo-guinan/bull-board/pull/317)
- chore(deps): bump path-parse from 1.0.6 to 1.0.7 [`#315`](https://github.com/leo-guinan/bull-board/pull/315)
- Added a filter input to the queue menu [`#313`](https://github.com/leo-guinan/bull-board/pull/313)
- fix: Allow bull-board to be bundled with webpack, fixes #310 [`#312`](https://github.com/leo-guinan/bull-board/pull/312)
- feat: add pause / resume on queue level [`#309`](https://github.com/leo-guinan/bull-board/pull/309)
- fix: encode job-id, closes #307 [`#308`](https://github.com/leo-guinan/bull-board/pull/308)
- Add confirmation modal on "dangerous" actions, closes #241 [`#302`](https://github.com/leo-guinan/bull-board/pull/302)
- fix: fix @bull-board/ui paths, closes #300 [`#301`](https://github.com/leo-guinan/bull-board/pull/301)
- fix: break long words in logs, data & error stack-trace, fixes #298 [`#299`](https://github.com/leo-guinan/bull-board/pull/299)
- fix: show failed job as "green" in active tab [`#296`](https://github.com/leo-guinan/bull-board/pull/296)
- fix: Show failed jobs as failed in "latest" tab [`#293`](https://github.com/leo-guinan/bull-board/pull/293)
- fix: compile typings with rest of the lib [`#292`](https://github.com/leo-guinan/bull-board/pull/292)
- fix: point global common typings to use dist [`#290`](https://github.com/leo-guinan/bull-board/pull/290)
- Add fav icon & change the logo to be more consistent between OS's [`#288`](https://github.com/leo-guinan/bull-board/pull/288)
- Feat: Add pagination [`#287`](https://github.com/leo-guinan/bull-board/pull/287)
- Add koa adapter [`#285`](https://github.com/leo-guinan/bull-board/pull/285)
- Move bull-board to use mono-repo structure [`#281`](https://github.com/leo-guinan/bull-board/pull/281)
- chore(deps): bump dns-packet from 1.3.1 to 1.3.4 [`#279`](https://github.com/leo-guinan/bull-board/pull/279)
- feat: add addQueue and removeQueue methods [`#278`](https://github.com/leo-guinan/bull-board/pull/278)
- fix ejs semicolon [`#275`](https://github.com/leo-guinan/bull-board/pull/275)
- fix: Update ioredis types & align adapter types, closes #271, #272 [`#273`](https://github.com/leo-guinan/bull-board/pull/273)
- chore(deps): bump lodash in /examples/with-multiple-instances [`#270`](https://github.com/leo-guinan/bull-board/pull/270)
- chore(deps): bump lodash from 4.17.20 to 4.17.21 in /examples/with-auth [`#269`](https://github.com/leo-guinan/bull-board/pull/269)
- refactor: extract common control code to queueProvider & jobProvider [`#268`](https://github.com/leo-guinan/bull-board/pull/268)
- Fix typo (bullMQAdapter to BullMQAdapter) [`#267`](https://github.com/leo-guinan/bull-board/pull/267)
- refactor: Remove client deps from lib's deps [`#263`](https://github.com/leo-guinan/bull-board/pull/263)
- V2 [`#261`](https://github.com/leo-guinan/bull-board/pull/261)
- Move tests to `tests` folder [`#260`](https://github.com/leo-guinan/bull-board/pull/260)

examples/with-multiple-instances/index.js
const { createBullBoard } = require('@leo-guinan/api');
const { BullMQAdapter } = require('@leo-guinan/api/bullMQAdapter');
const { ExpressAdapter } = require('@leo-guinan/express');
const { Queue: QueueMQ, Worker, QueueScheduler } = require('bullmq');
const express = require('express');

const sleep = (t) => new Promise((resolve) => setTimeout(resolve, t * 1000));

const redisOptions = {
  port: 6379,
  host: 'localhost',
  password: '',
  tls: false,
};

const createQueueMQ = (name) => new QueueMQ(name, { connection: redisOptions });

async function setupBullMQProcessor(queueName) {
  const queueScheduler = new QueueScheduler(queueName, {
    connection: redisOptions,
  });
  await queueScheduler.waitUntilReady();

  new Worker(queueName, async (job) => {
    for (let i = 0; i <= 100; i++) {
      await sleep(Math.random());
      await job.updateProgress(i);
      await job.log(`Processing job at interval ${i}`);

      if (Math.random() * 200 < 1) throw new Error(`Random error ${i}`);
    }

    return { jobId: `This is the return value of job (${job.id})` };
  });
}

const run = async () => {
  const exampleBullMq = createQueueMQ('BullMQ - instance1');
  const exampleBullMq2 = createQueueMQ('BullMQ - instance2');

  await setupBullMQProcessor(exampleBullMq.name);
  await setupBullMQProcessor(exampleBullMq2.name);

  const app = express();
  // Configure view engine to render EJS templates.
  app.set('views', __dirname + '/views');
  app.set('view engine', 'ejs');

  const serverAdapter1 = new ExpressAdapter();
  const serverAdapter2 = new ExpressAdapter();

  createBullBoard({
    queues: [new BullMQAdapter(exampleBullMq)],
    serverAdapter: serverAdapter1,
  });

  createBullBoard({
    queues: [new BullMQAdapter(exampleBullMq2)],
    serverAdapter: serverAdapter2,
  });

  serverAdapter1.setBasePath('/instance1');
  serverAdapter2.setBasePath('/instance2');

  app.use('/instance1', serverAdapter1.getRouter());
  app.use('/instance2', serverAdapter2.getRouter());

  app.use('/add', (req, res) => {
    const opts = req.query.opts || {};

    if (opts.delay) {
      opts.delay = +opts.delay * 1000; // delay must be a number
    }

    exampleBullMq.add('Add instance 1', { title: req.query.title }, opts);
    exampleBullMq2.add('Add instance 2', { title: req.query.title }, opts);

    res.json({
      ok: true,
    });
  });

  app.listen(3000, () => {
    console.log('Running on 3000...');
    console.log(
      'For the UI of instance1, open http://localhost:3000/instance1'
    );
    console.log(
      'For the UI of instance2, open http://localhost:3000/instance2'
    );
    console.log('Make sure Redis is running on port 6379 by default');
    console.log('To populate the queue, run:');
    console.log('  curl http://localhost:3000/add?title=Example');
    console.log('To populate the queue with custom options (opts), run:');
    console.log('  curl http://localhost:3000/add?title=Test&opts[delay]=9');
  });
};

// eslint-disable-next-line no-console
run().catch((e) => console.error(e));
